README
Business::CreditCard

These subroutines tell you whether a credit card number is self-consistent --
whether the last digit of the number is a valid checksum for the preceding
digits. 

Jon Orwant 

The Perl Journal and MIT Media Lab 

orwant@tpj.com 

Current maintainer is Ivan Kohler <ivan-business-creditcard@420.am>.
Please don't bother Jon with emails about this module.

Lee Lawrence <LeeL@aspin.co.uk> and Neale Banks <neale@lowendale.com.au>
contributed support for additional card types.  Lee also contributed a working
test.pl.



Changes
Revision history for Perl extension Business::CreditCard.

0.31  Mon Oct 19 18:51:35 PDT 2009
        - Add LICENSE section to POD documentation
        - Add META.yml to MANIFEST
        - Add Irish Laser card, thanks to Eoin Redmond for the heads-up.
        - Add documentation on what this module *is* and *is NOT* about.
        - Spelling fix in docs
        - correct misdocumentation of $Business::CreditCard::Country
        - Move test.pl to test/t and change to using Test::More, modify
          MANIFEST and Makefile.PL accordingly (thanks to Alexander Ciornii)
        - Silence unwanted warnings (thanks to Alexander Ciornii)
        - Discover cards starting with 644-649 are now recognized
        - Most Diner's Club cards (300-305, 3095 and 36) now processed as
          Discover
        - China Union pay now includes 624-626 and 628 in addition to 622 (and
          still identified as Discover outside China)
        - JCB identified as Discover in the US

0.30  Mon Dec 18 23:24:25 PST 2006
        - back after two and a half years; happy hanukkah!
	- added note about B:CC:Object
	- added new Discover 65 prefix
	- check for Switch before Visa as Switch has some BINs in ^4
        - accept masked numbers in cardtype()
        - add handling of card network peering arrangements, controllable via
          $Business::CreditCard::Country
	- identify Diner's club ^36 cards as MasterCard in US and Canada
        - identify China Union Pay cards as Discover cards outside China
        - identify China Union Pay cards

0.28  Thu Jul  1 01:17:32 PDT 2004
	- added Switch and Solo cards, patch from Max Becker
	- added proper Changes file
	- added documentation on cardtype() return values from
          tmurray@agronomy.org (closes: cpan #3330)

0.27  20 Jan 2002
	- small typo for amex cards

0.26  10 Jul 2001
0.25  10 Jul 2001
	- *sigh*

0.24  11 Jun 2001
	- added enRoute, JCB, BankCard, rewrote with regexes

0.23  3 May 2001
	- silly bug in test.pl

0.22  1 Feb 2001
	- new maintainer, MakeMaker installation

0.21  17 Jan 97
	- short numbers and numbers with letters are no longer kosher.

12 Jul 96 - created



test.pl
# Before `make install' is performed this script should be runnable with
# `make test'. After `make install' it should work as `perl test.pl'

######################### We start with some black magic to print on failure.

# Change 1..1 below to 1..last_test_to_print .
# (It may become useful if the test is moved to ./t subdirectory.)

BEGIN { $| = 1; print "1..2\n"; }
END {print "not ok 1\n" unless $loaded;}
use Business::CreditCard;
$loaded = 1;
print "ok 1\n";

######################### End of black magic.

# Insert your test code below (better if it prints "ok 13"
# (correspondingly "not ok 13") depending on the success of chunk 13
# of the test code):

#test 2
if( test_card_identification() ){ print "ok 2\n" }else{ print "not ok 2\n" }

sub test_card_identification{
        # 
        # For the curious the table of test number aren't real credit card
        # in fact they won't validate but they do obey the rule for the
        # cardtype table to identify the card type.
        #
        my %test_table=(
                '5212345678901234' =>   'MasterCard',
                '5512345678901234' =>   'MasterCard',
                '4123456789012' =>      'VISA card',
                '4512345678901234' =>   'VISA card',
                '341234567890123' =>    'American Express card',
                '371234567890123' =>    'American Express card',
                '30112345678901' =>     "Diner's Club/Carte Blanche",
                '30512345678901' =>     "Diner's Club/Carte Blanche",
                #'36123456789012' =>     "Diner's Club/Carte Blanche",
                '36123456789012' =>     'MasterCard',
                '38123456789012' =>     "Diner's Club/Carte Blanche",
                '201412345678901' =>    'enRoute',
                '214912345678901' =>    'enRoute',
                '6011123456789012' =>   'Discover card',
                '3123456789012345' =>   'JCB',
                '213112345678901' =>    'JCB',
                '180012345678901' =>    'JCB',
                '1800123456789012' =>   'Unknown',
                '312345678901234' =>    'Unknown',
                '4111xxxxxxxxxxxx' =>   'VISA card',
                '6599xxxxxxxxxxxx' =>   'Discover card',
                '6222xxxxxxxxxxxx' =>   'Discover card', #China Union Pay
        );
        while( my ($k, $v)=each(%test_table) ){
                if(cardtype($k) ne $v){
                        print "Card $k - should be $v cardtpe returns ",cardtype
($k),"\n";
                        return;
                }
        }
        return 1;
}



CreditCard.pm
package Business::CreditCard;

require Exporter;
use vars qw( @ISA $VERSION $Country );

@ISA = qw( Exporter );

$VERSION = "0.31";

$Country = 'US';

=head1 NAME

C<Business::CreditCard> - Validate/generate credit card checksums/names

=head1 SYNOPSIS

    use Business::CreditCard;
 
    print validate("5276 4400 6542 1319");
    print cardtype("5276 4400 6542 1319");
    print generate_last_digit("5276 4400 6542 131");

Business::CreditCard is available at a CPAN site near you.

=head1 DESCRIPTION

These subroutines tell you whether a credit card number is
self-consistent -- whether the last digit of the number is a valid
checksum for the preceding digits.  

The validate() subroutine returns 1 if the card number provided passes
the checksum test, and 0 otherwise.

The cardtype() subroutine returns a string containing the type of
card.  The list of possible return values is more comprehensive than it used
to be, but additions are still most welcome.

Possible return values are:

  VISA card
  MasterCard
  Discover card
  American Express card
  enRoute
  JCB
  BankCard
  Switch
  Solo
  China Union Pay
  Laser
  Unknown

"Not a credit card" is returned on obviously invalid data values.

Versions before 0.31 may also have returned "Diner's Club/Carte Blanche" (these
cards are now recognized as "Discover card").

As of 0.30, cardtype() will accept a partial card masked with "x", "X', ".",
"*" or "_".  Only the first 2-6 digits and the length are significant;
whitespace and dashes are removed.  To recognize just Visa, MasterCard and
Amex, you only need the first two digits; to recognize almost all cards
except some Switch cards, you need the first four digits, and to recognize
all cards including the remaining Switch cards, you need the first six
digits.

The generate_last_digit() subroutine computes and returns the last
digit of the card given the preceding digits.  With a 16-digit card,
you provide the first 15 digits; the subroutine returns the sixteenth.

This module does I<not> tell you whether the number is on an actual
card, only whether it might conceivably be on a real card.  To verify
whether a card is real, or whether it's been stolen, or to actually process
charges, you need a Merchant account.  See L<Business::OnlinePayment>.

These subroutines will also work if you provide the arguments
as numbers instead of strings, e.g. C<validate(5276440065421319)>.  

=head1 PROCESSING AGREEMENTS

Credit card issuers have recently been forming agreements to process cards on
other networks, in which one type of card is processed as another card type.

By default, Business::CreditCard returns the type the card should be treated as
in the US and Canada.  You can change this to return the type the card should
be treated as in a different country by setting
C<$Business::CreditCard::Country> to your two-letter country code.  This
is probably what you want to determine if you accept the card, or which
merchant agreement it is processed through.

You can also set C<$Business::CreditCard::Country> to a false value such
as the empty string to return the "base" card type.  This is probably only
useful for informational purposes when used along with the default type.

Here are the currently known agreements:

=over 4

=item Most Diner's club is now identified as Discover.  (This supercedes the earlier identification of some Diner's club cards as MasterCard inside the US and Canada.)

=item JCB cards in the 3528-3589 range are identified as Discover inside the US and Canada.

=item China Union Pay cards are identified as Discover cards outside China.

=back

=head1 NOTE ON INTENDED PURPOSE

This module is for verifying I<real world> B<credit cards>.  It is B<NOT> a
pedantic implementation of the ISO 7812 standard, a general-purpose LUHN
implementation, or intended for use with "creditcard-like account numbers".

=head1 AUTHOR

Jon Orwant

The Perl Journal and MIT Media Lab

orwant@tpj.com

Current maintainer is Ivan Kohler <ivan-business-creditcard@420.am>.
Please don't bother Jon with emails about this module.

Lee Lawrence <LeeL@aspin.co.uk>, Neale Banks <neale@lowendale.com.au> and
Max Becker <Max.Becker@firstgate.com> contributed support for additional card
types.  Lee also contributed a working test.pl.  Alexandr Ciornii
<alexchorny@gmail.com> contributed code cleanups.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 1995,1996,1997 Jon Orwant
Copyright (C) 2001-2006 Ivan Kohler
Copyright (C) 2007-2009 Freeside Internet Services, Inc.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

=head1 SEE ALSO

L<Business::CreditCard::Object> is a wrapper around Business::CreditCard
providing an OO interface.  Assistance integrating this into the base
Business::CreditCard distribution is welcome.

L<Business::OnlinePayment> is a framework for processing online payments
including modules for various payment gateways.

=cut

@EXPORT = qw(cardtype validate generate_last_digit);

sub cardtype {
    my ($number) = @_;

    $number =~ s/[\s\-]//go;
    $number =~ s/[x\*\.\_]/x/gio;

    return "Not a credit card" if $number =~ /[^\dx]/io;

    #$number =~ s/\D//g;
    {
      local $^W=0; #no warning at next line
      return "Not a credit card" unless length($number) >= 13 && 0+$number;
    }

    return "Switch"
      if $number =~ /^49(03(0[2-9]|3[5-9])|11(0[1-2]|7[4-9]|8[1-2])|36[0-9]{2})[\dx]{10}([\dx]{2,3})?$/o
      || $number =~ /^564182[\dx]{10}([\dx]{2,3})?$/o
      || $number =~ /^6(3(33[0-4][0-9])|759[0-9]{2})[\dx]{10}([\dx]{2,3})?$/o;

    return "VISA card" if $number =~ /^4[\dx]{12}([\dx]{3})?$/o;

    return "MasterCard"
      if   $number =~ /^5[1-5][\dx]{14}$/o
      ;# || ( $number =~ /^36[\dx]{12}/ && $Country =~ /^(US|CA)$/oi );

    return "Discover card"
      if   $number =~ /^30[0-5][\dx]{11}([\dx]{2})?$/o  #diner's: 300-305
      ||   $number =~ /^3095[\dx]{10}([\dx]{2})?$/o     #diner's: 3095
      ||   $number =~ /^3[68][\dx]{12}([\dx]{2})?$/o    #diner's: 36
      ||   $number =~ /^6011[\dx]{12}$/o
      ||   $number =~ /^64[4-9][\dx]{13}$/o
      ||   $number =~ /^65[\dx]{14}$/o
      || ( $number =~ /^62[24-68][\dx]{13}$/o && uc($Country) ne 'CN' ) #CUP
      || ( $number =~ /^35(2[89]|[3-8][\dx])[\dx]{10}$/o && uc($Country) eq 'US' );

    return "American Express card" if $number =~ /^3[47][\dx]{13}$/o;

    #return "Diner's Club/Carte Blanche"
    #  if $number =~ /^3(0[0-59]|[68][\dx])[\dx]{11}$/o;

    #"Diners Club enRoute"
    return "enRoute" if $number =~ /^2(014|149)[\dx]{11}$/o;

    return "JCB" if $number =~ /^(3[\dx]{4}|2131|1800)[\dx]{11}$/o;

    return "BankCard" if $number =~ /^56(10[\dx][\dx]|022[1-5])[\dx]{10}$/o;

    return "Solo"
      if $number =~ /^6(3(34[5-9][0-9])|767[0-9]{2})[\dx]{10}([\dx]{2,3})?$/o;

    return "China Union Pay"
      if $number =~ /^62[24-68][\dx]{13}$/o;

    return "Laser"
      if $number =~ /^6(304|7(06|09|71))[\dx]{12,15}$/o;

    return "Unknown";
}

sub generate_last_digit {
    my ($number) = @_;
    my ($i, $sum, $weight);

    $number =~ s/\D//g;

    for ($i = 0; $i < length($number); $i++) {
	$weight = substr($number, -1 * ($i + 1), 1) * (2 - ($i % 2));
	$sum += (($weight < 10) ? $weight : ($weight - 9));
    }

    return (10 - $sum % 10) % 10;
}

sub validate {
    my ($number) = @_;
    my ($i, $sum, $weight);
    
    return 0 if $number =~ /[^\d\s]/;

    $number =~ s/\D//g;

    return 0 unless length($number) >= 13 && 0+$number;

    for ($i = 0; $i < length($number) - 1; $i++) {
	$weight = substr($number, -1 * ($i + 2), 1) * (2 - ($i % 2));
	$sum += (($weight < 10) ? $weight : ($weight - 9));
    }

    return 1 if substr($number, -1) == (10 - $sum % 10) % 10;
    return 0;
}

1;


